### ISSUE TITLE
CAM left in unusable state after ca_request_close()

### ISSUE THREAD

**Issue by Jalle19** [catalinii/minisatip#1100]
When an incomplete write occurs, the CAM ends up being unusable until minisatip is restarted:
```
[21/04 22:12:51.331 AD7]: incomplete write to CA 7 fd 24, expected 5 got -1, errno 5 Input/output error
```

Since v1.2, `ca_request_close()` is called which just closes the CAM socket. Before, a proper reset was issued:

```
ioctl(fd, CA_RESET, &info)
```

**Comment by Jalle19** [comment]
@catalinii what's your opinion on changing this back to using CA_RESET?

**Comment by Yuri666** [comment]
i can confirm the issue
[log.txt.zip](https://github.com/catalinii/minisatip/files/11299059/log.txt.zip)

**Comment by catalinii** [comment]
Yeah changing back to uninitialized state and using CA_RESET should work

**Comment by Jalle19** [comment]
Turns out that closing the socket triggers the logic in the `ca_reconnect` poller, which calls `dvbca_init_dev()` which calls `ca_init_en50221` which does `ioctl(fd, CA_RESET, &info)`, so the issue doesn't seem to be that the device isn't getting reset properly. There's something else that causes minisatip to lose the ability to actually use the CAM for decryption after a reset, but I'm not sure where.

**Comment by catalinii** [comment]
Can you enable -v ca when that happens and post the log?

**Comment by Jalle19** [comment]
It happens exceedingly rarely, less than once a month for me. IIRC `-v ca` prints so much that I end up running out of disk space, but I'll see if I can enable it.

**Comment by Yuri666** [comment]
To reproduce this just needs pull and put card to CAM

**Comment by Jalle19** [comment]
Could be related to why things don't work after a CAM reset:

```
May 26 14:52:21 vladimir minisatip[51448]: [26/05 14:52:21.363 main]: ad 9 [9], another update_pids is in progress skipping
May 26 14:52:21 vladimir minisatip[51448]: [26/05 14:52:21.363 main]: CAPMT not found for pmt 11
May 26 14:52:21 vladimir minisatip[51448]: [26/05 14:52:21.363 main]: Skip deleting pmt for ddci adapter 9
```

Oddly this (incomplete write) happened to me today at almost the exact same time, and almost exactly 12 hours since minisatip was started (I restart it at 3 AM every night):

```
$ journalctl -u minisatip | grep 'incomplete write'
May 25 14:50:33 vladimir minisatip[47662]: [25/05 14:50:33.427 AD9]: incomplete write to CA 9 fd 28, expected 5 got -1, errno 5 Input/ou
May 26 14:51:49 vladimir minisatip[51448]: [26/05 14:51:49.803 AD9]: incomplete write to CA 9 fd 28, expected 5 got -1, errno 5 Input/ou
```

**Comment by Jalle19** [comment]
Actually no, the "another update_pids is in progress skipping" happens regularly, so probably not related

**Comment by Jalle19** [comment]
Aha, but `CAPMT not found for pmt X` only appears after a CAM reset

**Comment by Jalle19** [comment]
Trying this for the time being:
```javascript
const { spawn } = require('child_process')
const Journalctl = require('@seydx/journalctl')

const journalctl = new Journalctl({
  unit: 'minisatip',
})

journalctl.on('event', (event) => {
  const message = event['MESSAGE']

  if (message.includes('incomplete write')) {
    console.log(`Received offending log line, restarting: ${message}`)

    spawn('systemctl', ['restart', 'minisatip'])
  }
})
```

**Comment by Jalle19** [comment]
Only with general log level

**Comment by Jalle19** [comment]
[incomplete_write_20250325.log](https://github.com/user-attachments/files/19448257/incomplete_write_20250325.log)

Incomplete write occurs at `Mar 25 13:57:17` on AD8. After this, no channel will successfully open on AD8, only on AD9 (some channels are assigned there, most are not). In the end I give up and restart minisatip.

**Comment by catalinii** [comment]
The situation is more complex, when CA is closed, only the sockets are deleted.

What should happen is, DDCI close (to trigger this from ca. if ddci is enabled), pmts to be deleted from the CA adapter.

**Comment by Jalle19** [comment]
There are many functions, some of which do nothing:

```
int ddci_close_dev(adapter *ad) {
    if (ad->fe_sock > 0)
        sockets_del(ad->fe_sock);
    ad->fe_sock = -1;
    return TABLES_RESULT_OK;
}
SCA_op ddci;

void ddci_close_device(ddci_device_t *c) {}

int ddci_close_all() {
    int i;
    for (i = 0; i < MAX_ADAPTERS; i++)
        if (ddci_devices[i] && ddci_devices[i]->enabled) {
            ddci_close_device(ddci_devices[i]);
        }
    return 0;
}

int ddci_close(adapter *a) { return 0; }
```

Confusing...

**Comment by Jalle19** [comment]
@catalinii I'd appreciate some help on this, this bug is the only thing that makes me have to restart minisatip every now and then but I'm not familiar enough with the init/close cycle of all components/modules to know what to do

**Comment by Jalle19** [comment]
I can't find anything that reinitializes the DDCI device after we call `ddci_close_dev`.
